Overhaul LibationCli and add Download Quality Option

Mbucari opened the pull request:
## [Overhaul LibationCli](https://github.com/rmcrackan/Libation/pull/651/commits/29803c6ba0e659a0016de64654c1a68b9cfdc90a)
- Add version verb with option to check for upgrade
- Add Search verb to search the library
- Add export file type inference
- Add more set-status options
- Add console progress bar and ETA
- Add processable option to liberate specific book IDs
- Scan accounts by nickname or account ID
- Improve startup performance for help and on parsing error
- More useful error messages

Unfortunately [there's a bug in CommandLineParser](https://github.com/commandlineparser/commandline/issues/821) that necessitated rolling my own help text implementation, but it's not too convoluted.

## [Add setting to choose downloaded audio quality](https://github.com/rmcrackan/Libation/pull/651/commits/423b5312f7d3d75e515bca7c8fb890a9e86634c9) #648

rmcrackan commented:
btw I'm not ignoring you. A: There's been a lot of churn here so I want to give you time to get everything done that you want. B: I've had very little screen time today. I should have time tomorrow to give it the attention it deserves.

Mbucari commented:
> tomorrow

No rush, it's Sunday! I've reached a stopping point now. I Found a weird threading bug that I couldn't explain and spent about 2 hours trying to fix it. I still can't explain it, but it's gone now so who cares.

BTW, do you get an email every time I make a commit or do a forced push on an open PR? If so, sorry for the spam!

rmcrackan commented:
> do you get an email every time I make a commit or do a forced push on an open PR?

Yes. It's no problem though. Gmail is excellent about putting them in 1 thread whether it's 2 or 200.

rmcrackan commented on Source/LibationCli/LibationCli.csproj:
```diff
@@ -3,7 +3,8 @@
 	<PropertyGroup>
 		<OutputType>Exe</OutputType>
-		<TargetFramework>net7.0</TargetFramework>
+		<TargetFramework>net7.0-windows</TargetFramework>
```
Interesting. How will `net7.0-windows` jive with linux?

Mbucari replied:
It doesn't matter for releases because they're built with the publish command and the release identifier is overridden via the command line argument. This is just for debugging on windows machines. It's needed for debugging so that both WindowsConfigApp and LibationCli can be built with the same output directory.

rmcrackan commented on Source/LibationFileManager/AudibleFileStorage.cs:
```diff
@@ -126,7 +126,16 @@ protected override List<LongPath> GetFilePathsCustom(string productId)
             var regex = GetBookSearchRegex(productId);
-            return BookDirectoryFiles.FindFiles(regex);
+
+            //Find all extant files matching the priductID
```
Ah, the ever expanding complexity to save users from themselves. (btw -- typo: priductID)

Mbucari replied:
Yup! I Always keep the asin in my naming templates, but I had removed it when helping another user and discovered that I could not convert liberated m4bs to mp3 because it couldn't find them.

rmcrackan commented on Source/LibationCli/Options/_ProcessableOptionsBase.cs:
```diff
@@ -32,13 +43,28 @@ protected static TProcessable CreateProcessable<TProcessable>(EventHandler<Libra
 		}
 
-		protected static async Task RunAsync(Processable Processable)
+		protected async Task RunAsync(Processable Processable)
```
With a little refactoring, you might be able to speed up some of these loops with await foreach or Task.WhenAll.

Mbucari replied:
I've thought about adding a multithread feature to the GUI to allow users to run multiple processables simultaneously. It's something I'm interested in doing, so I'll save multithreading the cli until then. Please remind me then if I forget.

rmcrackan commented on Source/LibationCli/Options/SearchOptions.cs:
```diff
@@ -0,0 +1,55 @@
+namespace LibationCli.Options;
+
+[Verb("search", HelpText = "Search for books in your library")]
+internal class SearchOptions : OptionsBase
```
Interesting. I never considered that search might be useful from CLI.

Mbucari replied:
When I added options to specify asins for `liberate`, `convert`, and `set-status` the question of how are users going to find those asins naturally led to including a search.

rmcrackan commented on Source/LibationCli/Options/ExportOptions.cs:
```diff
@@ -29,26 +27,38 @@ class Options
 		*/
 		#endregion
-		[Option(shortName: 'x', longName: "xlsx", SetName = "xlsx", Required = true)]
+		[Option(shortName: 'x', longName: "xlsx", HelpText = "Microsoft Excel Spreadsheet", SetName = "Export Format")]
```
This uses `SetName` and SetDownloadStatusOptions uses `Group`. Are they the same? Similar?

Mbucari replied:
Simmilar. Both mean that the options are mutually exclusive, but at least one option in a `Group` is required and all options in a `Set` are optional.

I did screw this up though, and the three format options should be in different sets.

rmcrackan commented on Source/LibationCli/ConsoleProgressBar.cs:
```diff
@@ -0,0 +1,76 @@
+
+namespace LibationCli;
+
+internal class ConsoleProgressBar
```
CLI UI work is simultaneously frustrating and seductive. I remember [doing one years ago](https://github.com/rmcrackan/Dinah.Core/tree/master/Dinah.Core/ConsoleLib). No idea if it actually holds up lol

Mbucari replied:
Huh, didn't even know that was there. I wasn't trying to reinvent the wheel, but it was fun :)

rmcrackan replied:
Exactly. There was no point in me re-inventing the wheel either -- there are a ton of these -- except that it was fun

rmcrackan commented:
Done. I have some minor questions and I'm otherwise ready to merge when you are. This kind of backend work looks satisfying.

Mbucari commented on Source/LibationCli/Setup.cs:
```diff
@@ -1,21 +1,20 @@
 		{
+			//Determine variety by the dlls present in the current directory.
+			//Necessary to be able to check for upgrades.
+			var variety = System.IO.File.Exists("System.Windows.Forms.dll") ? Variety.Classic : Variety.Chardonnay;
```
Good idea.

Mbucari commented:
I've addressed all your comments. If you wouldn't mind holding off on a merge for a few hours, I'm going to see if I can't fix #652 as well. It looks like there may be a way to add tags with more than 4 chars to an m4b file, and I want to explore that.

rmcrackan commented:
> I've addressed all your comments. If you wouldn't mind holding off on a merge for a few hours, I'm going to see if I can't fix https://github.com/rmcrackan/Libation/issues/652 as well. It looks like there may be a way to add tags with more than 4 chars to an m4b file, and I want to explore that.

No problem. Just ping me here when you're ready. I'm done with the main review.

Mbucari commented:
All right, I'm ready for a review and merge on [Add audiobookshelf tags for m4b and mp3](https://github.com/rmcrackan/Libation/pull/651/commits/f2d475a9b05fd77287d31228ae80bfe548f8f3a9)

This should quell most of the audiobookshelf user's bitching for a while lol

rmcrackan commented:
Which tickets need to be closed? #652 and #648 ?

Mbucari commented:
> Which tickets need to be closed? #652 and #648 ?

Yup, just those 2.
